### PR TITLE
Add support for wasm

### PR DESCRIPTION
**Pull request by dmeijboom**
This PR adds support for wasm when using the `build.rs` flow (using the macro is untested and might not `work). Since `progenitor-client` uses the `stream` feature of `reqwest` it also depends on: https://github.com/seanmonstar/reqwest/issues/1704

### CODE REVIEW

**Comment by ahl** [comment]
This is neat. Thank you for doing this. Is it possible to do any testing of the wasm target in CI?

**Comment by dmeijboom** [comment]
> This is neat. Thank you for doing this. Is it possible to do any testing of the wasm target in CI?

Yes, definitely! However, it's blocked by support for the `stream` feature in Reqwest when using wasm. When that's merged I'll look at the tests.

File: progenitor-impl/src/lib.rs [hunk 1]
```diff
@@ -259,13 +259,18 @@ impl Generator {
...
                 ) -> Self {
-                    let dur = std::time::Duration::from_secs(15);
-                    let client = reqwest::ClientBuilder::new()
-                        .connect_timeout(dur)
-                        .timeout(dur)
-                        .build()
-                        .unwrap();
-                    Self::new_with_client(baseurl, client, #inner_value)
+                    #[cfg(target_arch = "wasm32")]
+                    let client = reqwest::ClientBuilder::new();
```
[the remark this answers is not in the document]
Comment (dmeijboom) [reply progenitor-impl/src/lib.rs:263]: Both `connect_timeout(...)` and `timeout(...)` aren't supported in wasm with reqwest.

File: progenitor-client/src/progenitor_client.rs [hunk 2]
```diff
@@ -76,6 +81,7 @@ impl<T: DeserializeOwned> ResponseValue<T> {
     }
 }
 
+#[cfg(not(target_arch = "wasm32"))]
```
[the remark this answers is not in the document]
Comment (dmeijboom) [reply progenitor-client/src/progenitor_client.rs:84]: The type `reqwest::Upgraded` doesn't exist in reqwest on wasm. So I assume it's not supported (yet?).

File: progenitor-client/src/progenitor_client.rs [hunk 3]
```diff
@@ -11,9 +11,14 @@ use futures_core::Stream;
 use reqwest::RequestBuilder;
 use serde::{de::DeserializeOwned, Serialize};
 
+#[cfg(not(target_arch = "wasm32"))]
 type InnerByteStream =
     std::pin::Pin<Box<dyn Stream<Item = reqwest::Result<Bytes>> + Send + Sync>>;
 
+#[cfg(target_arch = "wasm32")]
+type InnerByteStream =
+    std::pin::Pin<Box<dyn Stream<Item = reqwest::Result<Bytes>>>>;
```
[the remark this answers is not in the document]
Comment (dmeijboom) [reply progenitor-client/src/progenitor_client.rs:20]: I know that a `Future` translates to a `Promise` (https://rustwasm.github.io/wasm-bindgen/api/js_sys/struct.Promise.html) in wasm32 which is `!Send`. I can't find the rationale behind this but I assume it has something to do with the browser runtime which is single threaded (unless you are using web workers).

**Comment by ahl** [comment]
@dmeijboom looks like fixture data needs some love

**Comment by dmeijboom** [comment]
> @dmeijboom looks like fixture data needs some love

Yeah, indeed. I want to pick this MR up after `reqwest` supports the stream feature. It seems like it's moving in the right direction so I hope to continue working on this MR soon!

File: progenitor-client/src/progenitor_client.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (dmeijboom) [reply progenitor-client/src/progenitor_client.rs:15]: I removed `Send + Sync` entirely (also for non-wasm targets) and it seems to compile. Can it be safely removed or was there a reason for adding both markers to the `InnerByteStream` type?

**Comment by ahl** [comment]
this might be unblocked now by #316

**Comment by dmeijboom** [comment]
> this might be unblocked now by #316

Cool :) I'll update the MR later this week!

**Comment by dmeijboom** [comment]
Sorry for the delay 😅 I rebased the PR and updated the fixtures.

**Comment by dmeijboom** [comment]
> Is it possible to do any testing of the wasm target in CI?
> Some CI for the wasm target should now be possible.

I added a basic test case for wasm. It does nothing more than creating a client but at least we can guarantee that clients generated by progenitor are going to compile on wasm. Not sure where to go from here so I'm up for suggestions whether to add more test cases.

File: example-build/tests/wasm.rs [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+
+wasm_bindgen_test::wasm_bindgen_test_configure!(run_in_browser);
+
+include!(concat!(env!("OUT_DIR"), "/codegen.rs"));
+
+#[wasm_bindgen_test]
+fn test_client_new() {
+    let client = Client::new("http://foo/bar");
+    assert!(client.baseurl == "http://foo/bar");
+}
```
Comment (ahl) [inline example-build/tests/wasm.rs:11]: the purpose of the examples is for people to copy to them; let's keep the tests out of here.

File: progenitor-client/src/progenitor_client.rs [hunk 5]
```diff
@@ -11,6 +11,11 @@ use futures_core::Stream;
 use reqwest::RequestBuilder;
 use serde::{de::DeserializeOwned, Serialize};
 
+#[cfg(target_arch = "wasm32")]
+type InnerByteStream =
+    std::pin::Pin<Box<dyn Stream<Item = reqwest::Result<Bytes>>>>;
+
```
Comment (ahl) [inline progenitor-client/src/progenitor_client.rs:17]: let's move this below the non-wasm one

File: progenitor-impl/src/lib.rs [hunk 6]
```diff
@@ -345,13 +345,18 @@ impl Generator {
...
```
Comment (ahl) [inline progenitor-impl/src/lib.rs:349]: ditto; move this below non-wasm

**Review by ahl — COMMENTED** [review]
looking good; let's move that test and fix a couple of nits and it should be good to go

File: .github/workflows/rust.yml [hunk 7]
```diff
@@ -50,13 +50,20 @@ jobs:
         toolchain: nightly
         components: rustfmt
         default: false
+        target: wasm32-unknown-unknown
```
[the remark this answers is not in the document]
Comment (dmeijboom) [reply .github/workflows/rust.yml:74]: Yeah indeed, only Linux should be fine

File: example-build/tests/wasm.rs (hunk 4 above)
Comment (dmeijboom) [reply example-build/tests/wasm.rs]: I agree, moved to `example-wasm`!

**Comment by dmeijboom** [comment]
> looking good; let's move that test and fix a couple of nits and it should be good to go

Done! I created a new project called `example-wasm` as I don't want to pollute the other crates with wasm-bindgen-test stuff.

File: example-wasm/Cargo.toml [hunk 8]
```diff
@@ -0,0 +1,21 @@
+[package]
+name = "example-wasm"
+version = "0.1.0"
+edition = "2021"
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
```
Comment (ahl) [inline example-wasm/Cargo.toml:6]: delete

File: example-build/Cargo.toml [hunk 9]
```diff
@@ -9,7 +9,6 @@ chrono = { version = "0.4", features = ["serde"] }
 progenitor-client = { path = "../progenitor-client" }
 reqwest = { version = "0.11.16", features = ["json", "stream"] }
 base64 = "0.21"
-rand = "0.8"
```
Comment (ahl) [inline example-build/Cargo.toml:12]: let's leave it in? clients may need this... even if this one doesn't

**Review by ahl — CHANGES_REQUESTED** [review]
nit

**Comment by dmeijboom** [comment]
Also fixed the wasm tests! Forgot to setup chromedriver.

**Comment by ahl** [comment]
CI is still failing

**Comment by dmeijboom** [comment]
> CI is still failing

The `wasm-bindgen-test-runner` binary was missing. Should be fixed. There isn't an easy way to test GitHub actions unfortunately so I hope/think it works but wasn't able to validate it.

**Comment by jayvdb** [comment]
https://github.com/nektos/act is a nice tool for running github actions locally - often the workflows need a bit of fiddling to make it feasible to run locally.  I havent tried it yet with this project.

**Comment by ahl** [comment]
Let's gooo!

I'm excited to have this; thank you for the contribution. I'm optimistic about this being easy to maintain; and hope you'll be around to help us if we hit any snags.

**Pull request merged**